File: packages/js-sdk/src/template/templateApi.ts
import { ApiClient, components, handleApiError } from '../api'
import { ConnectionConfig, ConnectionOpts } from '../connectionConfig'
import { TemplateError } from '../errors'
import { Paginator } from '../paginator'

/**
 * Information about a sandbox template.
 */
export interface TemplateInfo {
  /**
   * Identifier of the template.
   */
  templateId: string

  /**
   * Identifier of the last successful build for the template.
   */
  buildId: string

  /**
   * Number of CPUs the template is configured with.
   */
  cpuCount: number

  /**
   * Amount of memory in MiB the template is configured with.
   */
  memoryMB: number

  /**
   * Disk size of the template in MiB.
   */
  diskSizeMB: number

  /**
   * Whether the template is public or only accessible by the team.
   */
  public: boolean

  /**
   * Aliases of the template.
   *
   * @deprecated Use {@link TemplateInfo.names} instead.
   */
  aliases: string[]

  /**
   * Names of the template (namespace/alias format when namespaced).
   */
  names: string[]

  /**
   * Time when the template was created.
   */
  createdAt: Date

  /**
   * Time when the template was last updated.
   */
  updatedAt: Date

  /**
   * Time when the template was last used, or `null` if it was never used.
   */
  lastSpawnedAt: Date | null

  /**
   * Number of times a sandbox was spawned from the template.
   */
  spawnCount: number

  /**
   * Number of times the template was built.
   */
  buildCount: number

  /**
   * Version of envd the template was built with.
   */
  envdVersion: string

  /**
   * User who created the template, or `null` if not available.
   */
  createdBy: components['schemas']['TeamUser'] | null

  /**
   * Status of the last build for the template.
   */
  buildStatus: components['schemas']['TemplateBuildStatus']
}

/**
 * Options for listing templates.
 */
export interface TemplateListOpts extends Omit<ConnectionOpts, 'signal'> {
  /**
   * Number of templates to return per page.
   *
   * @default 100
   */
  limit?: number

  /**
   * Token to the next page.
   */
  nextToken?: string
}

/**
 * Paginator for listing templates.
 *
 * @example
 * ```ts
 * const paginator = Template.list()
 * while (paginator.hasNext) {
 *   const templates = await paginator.nextItems()
 *   console.log(templates)
 * }
 * ```
 */
export class TemplatePaginator extends Paginator<TemplateInfo, ConnectionOpts> {
  constructor(opts?: TemplateListOpts) {
    super(opts, opts?.limit, opts?.nextToken)
  }

  async nextItems(opts?: ConnectionOpts): Promise<TemplateInfo[]> {
    // An exhausted paginator returns an empty list rather than throwing. The
    // sandbox and snapshot paginators currently throw here instead; they'll be
    // aligned to this behaviour.
    if (!this.hasNext) {
      return []
    }

    const config = new ConnectionConfig({ ...this.opts, ...opts })
    const client = new ApiClient(config)

    const res = await client.api.GET('/v2/templates', {
      params: {
        query: {
          limit: this.limit,
          nextToken: this.nextToken,
        },
      },
      signal: config.getSignal(opts?.requestTimeoutMs, opts?.signal),
    })

    const err = handleApiError(res, TemplateError)
    if (err) {
      throw err
    }

    this.updatePagination(res.response)

    return (res.data ?? []).map(
      ({
        templateID,
        buildID,
        createdAt,
        updatedAt,
        lastSpawnedAt,
        ...rest
      }: components['schemas']['Template']) => ({
        ...rest,
        templateId: templateID,
        buildId: buildID,
        createdAt: new Date(createdAt),
        updatedAt: new Date(updatedAt),
        lastSpawnedAt: lastSpawnedAt ? new Date(lastSpawnedAt) : null,
      })
    )
  }
}

File: packages/python-sdk/e2b/template/template_api.py
from dataclasses import dataclass
from datetime import datetime
from typing import List, Optional

from e2b.api.client.models import (
    Template as TemplateModel,
)
from e2b.api.client.models import (
    TeamUser,
    TemplateBuildStatus,
)


@dataclass
class TemplateInfo:
    """Information about a sandbox template."""

    template_id: str
    """Identifier of the template."""
    build_id: str
    """Identifier of the last successful build for the template."""
    cpu_count: int
    """Number of CPUs the template is configured with."""
    memory_mb: int
    """Amount of memory in MiB the template is configured with."""
    disk_size_mb: int
    """Disk size of the template in MiB."""
    public: bool
    """Whether the template is public or only accessible by the team."""
    aliases: List[str]
    """Aliases of the template."""
    names: List[str]
    """Names of the template (namespace/alias format when namespaced)."""
    created_at: datetime
    """Time when the template was created."""
    updated_at: datetime
    """Time when the template was last updated."""
    last_spawned_at: Optional[datetime]
    """Time when the template was last used, or None if it was never used."""
    spawn_count: int
    """Number of times a sandbox was spawned from the template."""
    build_count: int
    """Number of times the template was built."""
    envd_version: str
    """Version of envd the template was built with."""
    created_by: Optional[TeamUser]
    """User who created the template, or None if not available."""
    build_status: TemplateBuildStatus
    """Status of the last build for the template."""

    @classmethod
    def _from_template(cls, template: TemplateModel) -> "TemplateInfo":
        return cls(
            template_id=template.template_id,
            build_id=template.build_id,
            cpu_count=template.cpu_count,
            memory_mb=template.memory_mb,
            disk_size_mb=template.disk_size_mb,
            public=template.public,
            aliases=list(template.aliases) if template.aliases else [],
            names=list(template.names) if template.names else [],
            created_at=template.created_at,
            updated_at=template.updated_at,
            last_spawned_at=(
                template.last_spawned_at
                if isinstance(template.last_spawned_at, datetime)
                else None
            ),
            spawn_count=template.spawn_count,
            build_count=template.build_count,
            envd_version=template.envd_version,
            created_by=template.created_by,
            build_status=template.build_status,
        )
